File: 300-400/397_IntegerReplacement.py
class Solution:
    def integerReplacement(self, n: int) -> int:
        count = 0
        while n != 1:
            # 偶数(0bXX0)直接右移
            if (n & 1) == 0:
                n >>= 1
            else:
                #大于1的奇数格式为XX01或者XX11，对于前一种格式直接-1，对于后一种格式直接+1
                #单且仅当只有两位时候0b01(十进制的1)和0b11（十进制的3）时候，对于3，最少的步数应该是-1，因为3 - 1 = 2 / 2 = 1（三步）                #如果+1的话是3 + 1= 4 /2 =  2 / 2 = 1（四步），造成3这个数字比较特殊的原因是因为3的二进制只有2bit(0b11)
                n += -1 if (n & 2) == 0 or n == 3 else 1  
            count += 1
        return count

File: 300-400/397_IntegerReplacement.cc
#include <unordered_map>

using namespace std;

class Solution //记忆化搜索
{
private:
    unordered_map<int, int> memo;

public:
    int integerReplacement(int n)
    {
        if (n == 1)
        {
            return 0;
        }
        if (memo.count(n))
        {
            return memo[n];
        }
        if (n % 2 == 0)
        {
            return memo[n] = 1 + integerReplacement(n / 2);
        }
        return memo[n] = 2 + min(integerReplacement(n / 2), integerReplacement(n / 2 + 1));
    }
};

class Solution2 //枚举
{
public:
    int integerReplacement(int n)
    {
        if (n == 1)
        {
            return 0;
        }
        if (n % 2 == 0)
        {
            return 1 + integerReplacement(n / 2);
        }
        return 2 + min(integerReplacement(n / 2), integerReplacement(n / 2 + 1));
    }
};

class Solution3 //贪心
{
public:
    int integerReplacement(int n)
    {
        int ans = 0;
        while (n != 1)
        {
            if (n % 2 == 0)
            {
                ++ans;
                n /= 2;
            }
            else if (n % 4 == 1)
            {
                ans += 2;
                n /= 2;
            }
            else
            {
                if (n == 3)
                {
                    ans += 2;
                    n = 1;
                }
                else
                {
                    ans += 2;
                    n = n / 2 + 1;
                }
            }
        }
        return ans;
    }
};
